lib/reversetunnel/transport.go

```diff
@@ -88,12 +87,6 @@ func (t *TunnelAuthDialer) DialContext(ctx context.Context, _, _ string) (net.Co
 
 	addr, err := t.Resolver(ctx)
 	if err != nil {
-		if strings.Contains(err.Error(), "certificate is not trusted") {
-			err = trace.Wrap(
-				err,
-				"Your proxy certificate is not trusted or expired. Please update the certificate or follow this guide for self-signed certs: https://goteleport.com/docs/setup/admin/self-signed-certs",
-			)
-		}
```

Comment:
This was unnecessary and causing the message "Your proxy certificate..." to be printed multiple times when encountered, since we catch the error further down the callstack and wrap with the same message

Reply:
This also needs to be removed because it made a very subtle data race possible. This is because `t.Resolver` can be a cached resolver - in which case it may return the same `*trace.TraceErr` to two separate goroutines, and one may be reading `err.Error()` (which reads `*trace.TraceErr.Messages`) while the other calls `trace.Wrap(err, "Your proxy certificate....")` (which writes to `*trace.TraceErr.Messages`).

We create a caching resolver and use it to make a Dialer, which is then passed to `auth.NewClient` - and then the same dialer is used by both the gRPC client and HTTP client. The gRPC client spawns some goroutines - now the race condition is possible because the HTTP client can use the dialer at the same time that the gRPC client goroutine is using it.

See: https://github.com/gravitational/teleport/blob/23ee68caf87e78d897f95aa14ea0fe98764dcaf7/lib/auth/authclient/authclient.go#L84-L106